msal_extensions/token_cache.py

```diff
@@ -0,0 +1,196 @@
+        """
+        try:
+            return self._last_sync < os.path.getmtime(self._cache_location)
+        except IOError as exp:
+            if exp.errno != errno.ENOENT:
+                raise exp
+            return False
+
+    def _write(self):
+        """Handles actually committing the serialized form of this TokenCache to persisted storage.
+        For types derived of this, class that will be a file, which has the ability to track a last
+        modified time.
+        """
+        with open(self._cache_location, 'wb') as handle:
+            handle.write(self.serialize())
+
+    def _read(self):
+        """Fetches the contents of a file and invokes deserialization."""
+        with open(self._cache_location, 'rb') as handle:
+            contents = handle.read()
+        self.deserialize(contents)
+
+    def add(self, event, **kwargs):
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            super(FileTokenCache, self).add(event, **kwargs)  # pylint: disable=duplicate-code
+            self._write()
+
+    def modify(self, credential_type, old_entry, new_key_value_pairs=None):
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            super(FileTokenCache, self).modify(
+                credential_type,
+                old_entry,
+                new_key_value_pairs=new_key_value_pairs)
+            self._write()
+
+    def find(self, credential_type, **kwargs):  # pylint: disable=arguments-differ
+        with CrossPlatLock(self._lock_location):
+            if self._needs_refresh():
+                try:
+                    self._read()
+                except IOError as exp:
+                    if exp.errno != errno.ENOENT:
+                        raise exp
+            return super(FileTokenCache, self).find(credential_type, **kwargs)
+
+    def __getattr__(self, item):
+        # Instead of relying on implementers remembering to update _last_sync, just detect that one
+        # of the relevant methods has been called take care of it for derived types.
+        if item in ['_read', '_write']:
+            self._last_sync = int(time.time())
+        return super(FileTokenCache, self).__getattr__(item)  # pylint: disable=no-member
+
+
+class WindowsTokenCache(FileTokenCache):
+    """A SerializableTokenCache implementation which uses Win32 encryption APIs to protect your
+    tokens.
+    """
+    def __init__(self,
+                 cache_location=None,
+                 lock_location=None,
+                 entropy=''):
```

Comment:
I would suggest to NOT duplicate `cache_location` and `lock_location` here, because they are semantically NOT used by this sub-class anyway. The following approach would also be future-proof if the base class would invent some new parameters in the future.

```python
    def __init__(self, entropy="", **kwargs):
        super(..., self).__init__(**kwargs)
        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
```

Reply:
Fixed in 0327141

Reply:
Thanks. Just one more thing. You now have a typo of [duplicated line](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/15/files#diff-8e370721e03e7dd603efc5fe6fa2852aR160). :-)

Reply:
Derp! Fixed in e6009cdd5bdcc9c25282ee79392951b4114f0786